### PR TITLE
ES6 + parsing utils + migration update

### LINKED ISSUE #860: Add services to ES index mappings
### Summary
<table>
  <tr>
    <td><b>As a</b></td>
    <td>metadata content developer and user of OneStop</td>
  </tr>
  <tr>
    <td><b>I want to be able to</b></td>
    <td>
       quickly and easily locate services associated with a collection or granule metadata record,</td>
  </tr>
  <tr>
    <td><b>so that I can</b></td>
    <td>access the most up date information available for this dataset.</td>
  </tr>
</table>

### Description
Specifically, we want to support links coming from the ISO service metadata fields that NCEI has.

/gmi:MI_Metadata/gmd:identificationInfo/srv:SV_ServiceIdentification/srv:containsOperations/srv:SV_OperationMetadata/srv:connectPoint/gmd:CI_OnlineResource/.... 

SV_ServiceIdentification, containsOperations and connectPoint elements are repeatable.

Example: 
[Metadata record](https://ferret.pmel.noaa.gov/uafwaf/iso/granule/thredds_dodsC_ersstv3Agg.xml)

### Tasks
- [X] Review fields with Anna (metadata experts) determine what is stored
- [x] Update metadata parsing code to retrieve data from xPath and load into ElasticSearch
- [x] Add `serviceLinks` to collection and granule indices, schemas, and pipelines
- [x] Capture story on PSI board
- [ ] Update the [mapping documentation](https://github.com/cedardevs/onestop/blob/master/docs/design/iso-indexing-mapping.md#services) to indicate how services are now parsed

### LINKED ISSUE #866: Conditionally Toggle ES6/ES5 Nuances
### Summary
<table>
  <tr>
    <td><b>As a</b></td>
    <td>developer and deployer of OneStop</td>
  </tr>
  <tr>
    <td><b>I want to be able to</b></td>
    <td>
       quickly and easily configure OneStop to support a particular version or versions of Elasticsearch 
    </td>
  </tr>
  <tr>
    <td><b>So that I can</b></td>
    <td>manage multiple supported versions for flexibility moving forward, to provide better error handling, and use our resources more efficiently</td>
  </tr>
</table>

Support the different implementation details of es5 and es6 support via Spring Profiles
**Alternatively**, autodetect ES version within APIs and act accordingly? `client.getNode().getVersion()`

If autodetecting, we should probably just prevent the API from starting up if it's a version/range of Elasticsearch detected that we don't support.

#### ES6
- [x] client version
- [x] `_doc` instead of `doc` for populating index and search
- [x] [api-metadata] use the right "painless" script (both versions' scripts exist, but we need to ensure that a toggle works appropriately)
- [x] "transient setting [script.max_compilations_per_minute], not recognized"
- [x] [_all] is deprecated in 6.0+ and will be removed in 7.0. As a replacement, you can use [copy_to] on mapping fields to create your own catch all field."
- [x] error:[root_cause:[[type:illegal_argument_exception, reason:The provided expression [search_collection] matches an alias, specify the corresponding concrete indices instead.]], type:illegal_argument_exception, reason:The provided expression [search_collection] matches an alias, specify the corresponding concrete indices instead.]

### LINKED ISSUE #869: Update Developer Defaults to ES6
### Summary
<table>
  <tr>
    <td><b>As a</b></td>
    <td>developer of OneStop</td>
  </tr>
  <tr>
    <td><b>I want to be able to</b></td>
    <td>
       have my version of Elasticsearch default to the expected ES6 version
    </td>
  </tr>
  <tr>
    <td><b>So that I can</b></td>
    <td>align with our deployment configuration as well as know my local builds are consistent with other developers</td>
  </tr>
</table>

### Tasks
- [x] skaffold changes
- [x] helm chart changes
- [x] search onestop project for image/version usage and keep consistent

### LINKED ISSUE #897: Metadata parsing, validation, and loading consistency
### Summary
<table>
  <tr>
    <td><b>As a</b></td>
    <td>developer and operator of OneStop</td>
  </tr>
  <tr>
    <td><b>I want to be able to</b></td>
    <td>
       use a shared library to parse, analyze, and load metadata to the OneStop index in a consistent way with either http or kafka</td>
  </tr>
  <tr>
    <td><b>so that I can</b></td>
    <td>reduce code duplication, and assumptions about migrating from loading via http to kafka</td>
  </tr>
</table>

There are two parts to this story- 

The first is to use the shared library from the [schemas repo](https://github.com/cedardevs/schemas) for parsing and analyzing xml to make a staging index ready doc in api-metadata. 

The second part stems from the first part. Because the discovery schema is not enough for the staging index, we have to make the xml upload logic follow a similar step the kafka consumer logic follows to generate gcmd keywords, among other things. 

While we are in the moving code and simplifying, we want to maintain/implement the following rules for updating a record. This rules will make the migration from the XML Upload API to PSI much less painful for our operators. 
Ensure the following behavior- 
- [x] XML uploaded via the endpoint is parsed using shared libraries and existing logic about fileIds is maintained (disallow null identifiers and ambiguous records). 
- [x] a ParsedRecord coming from PSI with an PSI ID should be updated if that ID already exists in the index (existing behavior)
- [x] If a ParsedRecord comes from PSI with an ID that *is not* already in the index, check to see if that fileIdentifier or DOI already exists in the index. 
- [x] If there is a doc with that fileId or DOI, delete it from staging and search, and load it again from the ParsedRecord we got from PSI using the PSI ID (biggest change). 
- [x] If we get more than one doc, send back CONFLICT as we do now. 
- [x] All parsedRecords should be validated before going into the staging index
- [x] If we re-key a collection, make sure all associated granules internalParentIdentifier is updated in search. May require deleting all associated granules out of search.

Also behavior that might be easy to implement during the course of this work-
- [x] a valid record can have a fileIdentifier, a DOI, or both
- [ ] If PSI sends a tombstone, don't invalidate it, delete the record.

### PR DESCRIPTION
**Pull request by StrayCat1**
This pull request closes several stories -  
Closes #860 - services
Closes #897 - parsing utils, loading consistency, migration update
Closes #866 - ES6 toggle
Closes #869 -  ES6 dev default

For Specific changes and features related to #860 and #897, see the comments in this story -  
https://app.zenhub.com/workspaces/onestop-5caf5b03e94905314d5883a8/issues/cedardevs/onestop/897

### CODE REVIEW

File: api-metadata/src/main/groovy/org.cedar.onestop.api.metadata/service/InventoryManagerToOneStopUtil.groovy [hunk 1]
```diff
@@ -12,66 +17,105 @@ import static org.cedar.schemas.avro.psi.ValidDescriptor.*
...
     if (temporal == null) {
-      failures.add("Missing temporal analysis")
+      details << "Missing temporal analysis"
     }
-    if (identification && !identification?.fileIdentifierExists) {
-      failures.add('Missing fileIdentifier')
+    if (identification && (!identification?.fileIdentifierExists && !identification?.doiExists)) {
+      details << "Missing identifier - record contains neither a fileIdentifier nor a DOI"
+    }
+    if (identification && (!identification?.fileIdentifierExists && !identification?.doiExists)) {
```
[the remark this answers is not in the document]
Comment (StrayCat1) [reply api-metadata/src/main/groovy/org.cedar.onestop.api.metadata/service/InventoryManagerToOneStopUtil.groovy]: Good catch, will fix

[the remark this answers is not in the document]
Comment (StrayCat1) [reply api-metadata/src/main/groovy/org.cedar.onestop.api.metadata/service/InventoryManagerToOneStopUtil.groovy]: fix - https://github.com/cedardevs/onestop/commit/3acd33f971d76ae18e1f726dec235254f78ede07

File: api-metadata/src/main/groovy/org.cedar.onestop.api.metadata/service/MetadataManagementService.groovy [hunk 2]
```diff
@@ -60,132 +44,151 @@ class MetadataManagementService {
...
-                detail: "The identifiers in this document match more than one existing document. " +
-                    "Please GET records with ids [ ${existingIds.join(',')} ] and DELETE any " +
-                    "erroneous records. Ambiguity because of fileIdentifier [ ${fileId} ] and/or DOI [ ${doi} ]."
-            ]
+          results << result
+          return
+        }
+
+        String esId = id ?: null
+        if(existingIds?.size() == 1){ //this is an update
```
[the remark this answers is not in the document]
Comment (StrayCat1) [reply api-metadata/src/main/groovy/org.cedar.onestop.api.metadata/service/MetadataManagementService.groovy]: how bout this - 
https://github.com/cedardevs/onestop/commit/a61764457461360722d134b48a1a98345ffd01e2

File: elastic-common/src/main/groovy/org/cedar/onestop/elastic/common/DocumentUtil.groovy [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+    static String getFileIdentifierFromSource(Map source) {
+        return source.fileIdentifier
+    }
+
+    static String getParentIdentifier(Map document) {
+        return getSource(document).parentIdentifier
+    }
+
+    static String getParentIdentifierFromSource(Map source) {
```
[the remark this answers is not in the document]
Comment (forgo) [reply elastic-common/src/main/groovy/org/cedar/onestop/elastic/common/DocumentUtil.groovy]: @zebdelk  The concern i think was that there was no "parsedResponse" or "document" in scope where these getters were being used

[the remark this answers is not in the document]
Comment (zebdelk) [reply elastic-common/src/main/groovy/org/cedar/onestop/elastic/common/DocumentUtil.groovy]: The case I remember seeing in the group review the code was doing something like
def source = getSource(document)
getParentIdentifierFromSource(source)
But I haven't looked through to see if that's consistently true.

**Pull request merged**